### DISCUSSION TITLE
I'm facing build failures in Next.js due to the public folder's large bundle size being included.

### DISCUSSION THREAD

**Discussion by mirhamasala** [vercel/next.js#68160] [Help]
### Summary

We are experiencing an issue where our builds are consistently failing due to the bundle size being too large. After looking into it, I found that the public folder, which contains many images, is being included in our Function’s bundle, leading to this problem.

The main issue is that Next.js is bundling our entire root directory, including the public folder, which results in unnecessary files being included in the standalone output. This significantly increases the bundle size and causes build failures.

I am seeking guidance on why Next.js is bundling the entire root directory and if there is a better way to prevent the public folder from being included in the bundle. While I have a temporary workaround, I am looking for a more robust and permanent solution to this issue.

To temporarily resolve this, I have added the following configuration to our next.config.js:

```
outputFileTracingIncludes: {
  '/': ['src/content/**'],
},
outputFileTracingExcludes: {
  '/': [
    '.git/**',
    '.github/**',
    '.next/cache/**',
    '.vscode/**',
    'cypress/**',
    '.netlify/**',
    'public/**',
    'scripts/**',
    'src/app/**',
  ],
},
```

### Example

https://github.com/FilecoinFoundationWeb/filecoin-foundation

**Comment by Ahsan2001** [comment]
Step 1: Update your next.config.ts to this file

import type { NextConfig } from 'next'; const nextConfig: NextConfig = { productionBrowserSourceMaps: false, output: 'standalone', }; export default nextConfig;

Step 2: install this package npm install --save-dev rimraf

Step 3: replace you build script in package json "build": "next build && rimraf .next/cache",

**Comment by Corey-Code** [comment]
You're running into a common issue with Next.js when deploying serverless functions (e.g. on Vercel, Netlify, AWS Lambda) using output: 'standalone' mode — Next.js includes more files than necessary, including large assets from public/, which can bloat your bundle and exceed deployment limits.

Let’s walk through why this happens, what your workaround does, and how to implement a cleaner, robust fix.

**Why Is public/ Being Bundled?**
1) When output: 'standalone' is enabled, Next.js tries to trace all used files via @vercel/nft to ensure the resulting bundle has everything needed at runtime.

2) It over-approximates dependencies, sometimes pulling in unused files — especially if paths are dynamically resolved or non-standard libraries are used to access assets.

By default:
- public/ is not required for server-side code and should not be bundled.
- But if any server-side file references public/, even indirectly, it may get pulled in.

I hope this helps you find a resolution if not already.

**Reply by versatile15** [reply]
Thanks for the detailed explanation — this clarifies a lot. I hadn't realized how aggressively @vercel/nft traces dependencies when output: 'standalone' is enabled. It makes sense now why public assets are unintentionally pulled into the server bundle, especially with dynamic or indirect references.

In my case, it looks like a server-side utility was resolving paths relative to public/, which inadvertently triggered the inclusion. I’ll refactor to avoid referencing public/ from server code and will also look into excluding unnecessary files post-build if needed.

Appreciate the insight and the clean breakdown of what’s happening under the hood.

**Reply by gcb** [reply]
How can i extract more information when it's not obvious what is triggering this?

```
Turbopack build encountered 1 warnings:
./next.config.ts
Encountered unexpected file in NFT list
A file was traced that indicates that the whole project was traced unintentionally. Somewhere in the import trace below, there are:
- filesystem operations (like path.join, path.resolve or fs.readFile), or
- very dynamic requires (like require('./' + foo)).
To resolve this, you can
- remove them if possible, or
- only use them in development, or
- make sure they are statically scoped to some subfolder: path.join(process.cwd(), 'data', bar), or
- add ignore comments: path.join(/*turbopackIgnore: true*/ process.cwd(), bar)

Import trace:
  App Route:
    ./next.config.ts
    ./lib/generated_prisma/index.js
    ./lib/prisma.ts
    ./app/file/[file_id]/route.ts
```

route.ts above will load a file based on query parameters and serve it. There is not file path or fs operation on that file, at all. There are not even path strings on that code. It is all one library down, which does have a few
```
# lib/file_io.ts
const file_path = await getFilePathFromDB(file_id)
const file_content = await fsPromises.readFile(/*turbopackIgnore: true*/ file_path);
```

**Reply by gcb** [reply]
since this only happens with turbopack, i'm back using webpack. which is awfull because latest nextjs have tons of bugs now in dev mode with webpack. i guess the devs only test turbopack now.

**Reply by Firgrep** [reply]
Getting the same issue here. Using older prisma v6 (since v7 does not work for mongodb). I had a db package that contained multiple prisma clients and when imported via npm it worked fine, but putting it into a monorepo caused issues. I'm having to copy the contents in the dockerfile as a separate step. The warnings during build still happen but the app works. 

```
RUN for dir in /app/lib/db/prisma-client-*; do \
      name=$(basename "$dir"); \
      mkdir -p "./$name"; \
      cp "$dir/schema.prisma" "./$name/"; \
      cp "$dir"/libquery_engine-*.so.node "./$name/"; \
    done
```

**Comment by oeyoews** [comment]
A file was traced that indicates that the whole project was traced unintentionally. Somewhere in the import trace below, there are:xxx

**Comment by barry166** [comment]
The thing to debug is the trace, not the size of `public/` itself.

In `output: 'standalone'`, Next copies the files listed in the generated `.nft.json` traces. `public/` is not normally copied by the standalone writer, so if the whole folder appears in the function bundle it usually means some server-side import/path caused `@vercel/nft` to trace too broadly. Check the relevant trace files first:

```bash
find .next -name '*.nft.json' -print0 | xargs -0 grep -n 'public/'
find .next -name '*.nft.json' -print0 | xargs -0 grep -n 'FilecoinFoundationWeb'
```

Then fix the import that pulls the project root into the server trace, or keep a route-specific exclude as a last resort. Also make the keys route globs (`'/*'`, `'/api/*'`, etc.); using `'/'` only targets the root route and is easy to misunderstand.

**Comment by oeyoews** [comment]
When implementing logic involving local file paths within Next.js API routes—for instance, building an Electron app—there is no way around syntax like path.join or process.cwd unless I offload such operations via inter-process communication with the Electron main process.
